src/test/java/com/solbeg/MainTest.java

```diff
@@ -0,0 +1,231 @@
+
+        assertEquals(Users.generateUser(), Main.getOrGenerateUser(Optional::empty));
+    }
+
+    @Test
+    void ifUserNotProvided_shouldTrowNoSuchElementException() {
+
+        UserProvider userProvider = Optional::empty;
+
+        assertThrows(NoSuchElementException.class, () -> Main.retrieveBalance(userProvider));
+    }
+
+    @Test
+    void ifUserProvided_shouldReturnNotNullValue() {
+
+        UserProvider userProvider = new UserProvider() {
+            @Override
+            public Optional<User> getUser() {
+                return Optional.ofNullable(Users.generateUser());
+            }
+        };
+
+        assertNotNull(Main.retrieveBalance(userProvider));
+    }
+
+    @Test
+    void ifUserNotProvided_shouldThrowRuntimeExceptionWithMessage() {
+
+        UserProvider userProvider = Optional::empty;
+
+        assertThrowsExactly(RuntimeException.class, () -> Main.getUser(userProvider), "No user provided");
+    }
+
+    @Test
+    void ifCreditBalanceIsPresent_shouldReturnNotNull() {
+
+        UserBankAccountProvider userBankAccountProvider = new UserBankAccountProvider() {
+            @Override
+            public Optional<UserBankAccount> getUserBankAccount() {
+                return Optional.of(new UserBankAccount());
+            }
+        };
+
+        assertNotNull(Main.retrieveCreditBalance(userBankAccountProvider));
+    }
+
+    @Test
+    void ifUserWithEndOfEMailIsPresent_shouldReturnThisUser() {
+
+        UserProvider user = () -> Optional.ofNullable(Users.generateUser());
+
+        assertNotNull(Main.retrieveUserGmail(user));
+        assertEquals("m@gmail.com", Main.retrieveUserGmail(user).get().getEmail());
+    }
+
+    @Test
+    void ifUserNotProvidedFromAllProvider_shouldThrowNoSuchElementException() {
+
+        UserProvider userProvider = Optional::empty;
+        UserProvider userProviderFallBack = Optional::empty;
+
+        assertThrows(NoSuchElementException.class, () -> Main.getUserWithFallback(userProvider, userProviderFallBack));
+    }
+
+    @Test
+    void ifUserNotProvidedFromUserProvider_userShouldBeProvideFromFallBackProvider() {
+
+        UserProvider userProvider = Optional::empty;
+        UserProvider userProviderFallBack = new UserProvider() {
+            @Override
+            public Optional<User> getUser() {
+                return Optional.ofNullable(Users.generateUser());
+            }
+        };
+
+        assertNotNull(Main.getUserWithFallback(userProvider, userProviderFallBack));
+    }
+
+    @Test
+    void ifInputListIsEmpty_shouldThrowNoSuchElementException() {
+
+        assertThrows(NoSuchElementException.class, () -> Main.getUserWithMaxBalance(List.of()));
+    }
+
+    @Test
+    void shouldReturnCorrectBalance() {
+
+        OptionalDouble falseBalance = bankAccounts.stream()
+                .mapToDouble(bankAccount -> bankAccount.getCreditBalance().get().doubleValue())
+                .max();
+        OptionalDouble expectedBalance = bankAccounts.stream()
```

Comment:
do not write logic for summing the values
because you are testing the same logic))
then how you will know that the method is working properly

Reply:
if you need this false and expected balance then calculate it by calculator

Reply:
fixed